spec/lib/change_history_modifier_spec.rb

```diff
@@ -0,0 +1,149 @@
+    end
+  end
+
+  describe "#edit" do
+    let(:id_to_edit) { SecureRandom.uuid }
+    let(:change_history_entry) do
+      [
+        {
+          id: SecureRandom.uuid,
+          public_timestamp: "2020-02-20T18:00:00+00:00",
+          note: "Test 1",
+        }, {
+          id: id_to_edit,
+          public_timestamp: "2020-02-20T17:00:00+00:00",
+          note: "Test 2",
+        }, {
+          id: SecureRandom.uuid,
+          public_timestamp: "2020-02-20T16:00:00+00:00",
+          note: "Test 3",
+        }
+      ]
+    end
+    let(:first_edition) do
+      create(:edition,
+             current: false,
+             first_published_at: "2020-02-20 12:00:00",
+             locale: "en")
+    end
+    let(:second_edition) do
+      create(:edition,
+             document: first_edition.document,
+             number: 2,
+             locale: "en",
+             change_history: change_history_entry)
+    end
+
+    it "changes a specific change history entry note" do
+      expected_change_history = change_history_entry.dup
+
+      stub_publishing_api_put_content(second_edition.document.content_id, {})
+
+      change_history_modifier = described_class.new(second_edition.document.content_id, nil, nil)
+      change_history_modifier.edit(id_to_edit, "Test 99")
+
+      expected_change_history[1]["note"] = "Test 99"
+      expect(second_edition.reload.revision.metadata_revision.change_history)
+        .to eq(expected_change_history.map(&:with_indifferent_access))
+    end
+
+    it "raises an error when the change history ID does not exist" do
+      expect { described_class.new(second_edition.document.content_id, nil, nil).edit("abc123", "Test 99") }
+        .to raise_error("No change history entry with id abc123")
+    end
+  end
+
+  describe "#add" do
+    let(:change_history_entry) do
+      [
+        {
+          id: SecureRandom.uuid,
+          public_timestamp: "2020-02-20T16:00:00+00:00",
+          note: "Test 1",
+        },
+      ]
+    end
+    let(:first_edition) do
+      create(:edition,
+             current: false,
+             first_published_at: "2020-02-20 12:00:00",
+             locale: "en")
+    end
+    let(:second_edition) do
+      create(:edition,
+             document: first_edition.document,
+             number: 2,
+             locale: "en",
+             change_history: change_history_entry)
+    end
+
+    it "add a change history entry note" do
+      stub_publishing_api_put_content(second_edition.document.content_id, {})
+
+      change_history_modifier = described_class.new(second_edition.document.content_id, nil, nil)
+      change_history_modifier.add("Test 99", "2020-10-22 09:30")
+
+      expect(second_edition.reload.revision.metadata_revision.change_history.first["note"])
+        .to eq("Test 99")
+      expect(second_edition.revision.metadata_revision.change_history.first["public_timestamp"])
+        .to eq("2020-10-22T09:30:00+01:00")
+    end
+  end
```

Comment:
It looks like we're missing tests for the other responsibilities this class has: 

- checking edition is editable
- raising if edition is not changed
- calling edit draft edition service
- calling preview service

It will be a bit of a pain to test these based on the suggestion I gave for how to write the class 🤔I'm just going to have a ponder about that.

Reply:
I believe these are all covered now.  As always - thanks very much for your comments and suggestions.